### PR TITLE
SONARPY-453 Remove duplicate representation of docstring

### PR DESCRIPTION
**Pull request by guillaume-dequenne-sonarsource**

### CODE REVIEW

File: python-squid/src/test/java/org/sonar/python/DocstringExtractorTest.java [hunk 1]
```diff
@@ -74,19 +74,22 @@ private Token getDocstring(PythonGrammar nodeType, int line) {
 
     @Override
     public void scanFile(PythonVisitorContext visitorContext) {
-      scanNode(visitorContext.rootAstNode());
-
+      scanTree(visitorContext.rootTree());
```
Comment (benzonico) [inline python-squid/src/test/java/org/sonar/python/DocstringExtractorTest.java]: I don't think we need a separate method here. This scanTree method can be inlined I believe.

File: python-squid/src/test/java/org/sonar/python/tree/PythonTreeMakerTest.java [hunk 2]
```diff
@@ -726,61 +727,61 @@ public void nonlocalStatement() {
...
-
-    functionDefTree = parse("def func(x): pass", treeMaker::funcDefStatement);
-    assertThat(functionDefTree.parameters().all()).hasSize(1);
-
-    functionDefTree = parse("async def func(x): pass", treeMaker::funcDefStatement);
-    assertThat(functionDefTree.asyncKeyword().value()).isEqualTo("async");
-
-    functionDefTree = parse("def func(x) -> string : pass", treeMaker::funcDefStatement);
-    TypeAnnotation returnType = functionDefTree.returnTypeAnnotation();
+    FunctionDef functionDef = treeMaker.funcDefStatement(astNode);
```
Comment (benzonico) [inline python-squid/src/test/java/org/sonar/python/tree/PythonTreeMakerTest.java]: This is great to rename the variable, but, as this is unrelated to the task at hand could you try to keep it in a separate commit next time as it makes things hard to review : this is tough to separate refactoring from required change at review time. (no action required for this comment).

Comment (guillaume-dequenne-sonarsource) [reply python-squid/src/test/java/org/sonar/python/tree/PythonTreeMakerTest.java]: Indeed! Will make separate commits next time :)

File: python-squid/src/test/java/org/sonar/python/DocstringExtractorTest.java (hunk 1 above)
Comment (guillaume-dequenne-sonarsource) [reply python-squid/src/test/java/org/sonar/python/DocstringExtractorTest.java]: Tbh I'd like to keep it that way because scanTree recursively calls itself. If I inline, the benefit of getting rid of the method would be offset by the fact I'd need to reimplement the recursion through a loop/stack construct.

File: python-squid/src/main/java/org/sonar/python/DocstringExtractor.java [hunk 3]
```diff
@@ -36,48 +37,12 @@
...
-  private static Token extractDocstringFromSimpleStmt(@Nullable AstNode firstSimpleStmt) {
-    if (firstSimpleStmt != null) {
-      Token token = firstSimpleStmt.getToken();
-      if (token.getType().equals(PythonTokenType.STRING)) {
-        return token;
-      }
+  public static StringLiteral extractDocstring(@Nullable StatementList statements) {
+    if (statements != null) {
+      Statement firstStatement = statements.statements().get(0);
+      return firstStatement.is(Tree.Kind.EXPRESSION_STMT) && ((ExpressionStatement) firstStatement).expressions().size() == 1 &&
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-squid/src/main/java/org/sonar/python/DocstringExtractor.java]: Most importantly, I think I was missing a test case for that coverage, which I added. I also refactored the conditional expression as an if statement :)

File: python-squid/src/test/java/org/sonar/python/DocstringExtractorTest.java (hunk 1 above)
Comment (benzonico) [reply python-squid/src/test/java/org/sonar/python/DocstringExtractorTest.java]: missed the recursion, fair enough. But then I wonder why not make this a PythonCheckTree to avoid to reimplement the visit ?

Comment (guillaume-dequenne-sonarsource) [reply python-squid/src/test/java/org/sonar/python/DocstringExtractorTest.java]: I reproduced the previous implementation based on AstNode without questioning it... Sure it makes more sense to do it that way now.
